dataprep/clean/clean_country.py

```diff
@@ -0,0 +1,328 @@
+        x = x.astype(str).str.lower()
+        x = x.str.strip()
+        return x.apply(check_country, args=(input_format, False, strict))
+
+    x = str(x).lower()
+    x = x.strip()
+    return check_country(x, input_format, False, strict)
+
+
+def format_country(
+    row: pd.Series,
+    col: str,
+    input_format: str,
+    output_format: str,
+    fuzzy_dist: int,
+    errors: str,
+) -> pd.Series:
+    """
+    Function to transform a country instance into the
+    desired format
+    """
+    # pylint: disable=too-many-arguments
+    # check_country parses the value in row[col], and will return the index of the country
+    # in the DATA dataframe. The returned value "status" can be either "null"
+    # (which means row[col] contains a null value), "unknown" (in which case the value
+    # in row[col] could not be parsed) or "success" (a successful parse of the value).
+
+    country = str(row[col])
+    country = country.lower().strip()
+    result_index, status = check_country(country, input_format, True)
+
+    if fuzzy_dist > 0 and status == "unknown":
+        result_index, status = check_fuzzy_dist(country, input_format, fuzzy_dist)
+
+    if status == "null":
+        STATS["null"] += 1
+        row[f"{col}_clean"] = np.nan
+        return row
+    if status == "unknown":
+        if errors == "raise":
+            raise ValueError(f"unable to parse value {row[col]}")
+        STATS["unknown"] += 1
+        row[f"{col}_clean"] = row[col] if errors == "ignore" else np.nan
+        return row
+
+    result = DATA.iloc[result_index][output_format]
+    row[f"{col}_clean"] = str(int(result)) if output_format == "numeric" else result
+    if row[col] != row[f"{col}_clean"]:
+        STATS["cleaned"] += 1
+    return row
+
+
+@lru_cache(maxsize=2 ** 20)
+def check_country(country: str, input_format: str, clean: bool, strict: bool = False) -> Any:
+    """
+    Finds the index of the given country in the DATA dataframe.
+
+    Parameters
+    ----------
+    country
+        pandas Series of countries or str/int country value
+    input_format
+        the ISO 3166 input format of the country
+    clean
+        If True the function returns True/False to be used by the validate country function.
+        If False, a tuple (index, status) is returned.
+    strict
+        If True, the regex match for input types "name" and "offical" must occur at the
+        beginning of the input. If False, a match can be anywhere within the input.
+    """
+    if country in NULL_VALUES:
+        result = ("", "null") if clean else False
+        return result
+
+    if input_format == "auto":
+        input_format = get_format_from_name(country)
+
+    if input_format in ("regex", "name", "official"):
+        for index, regex in enumerate(REGEXES):
+            if not strict and regex.search(country):
+                result = (index, "success") if clean else True
+                return result
+            if strict and regex.match(country):
```

Comment:
Good point, I agree using re.match here doesn't make sense. I think it could be worth considering not using regexes at all when `strict = True` and instead look for a direct match with the country name in the given input format. `re.fullmatch` would match some inputs we want for example:

"italia" -> "Italy"
"bosnia" -> "Bosnia and Herzegovina"

But it would also match some things we might not want, for example:

The regex for "Ireland" is "^(?!.\*north.\*).*ireland" and the regex for "Christmas Island" is "christmas"

"country: ireland" -> "Ireland"
"christmas" -> "Christmas Island"

I think not using regexes here would simplify the behaviour and keep the matching more strict but it also reduces the functionality slightly.

Reply:
Great point, I think you're right that looking for a direct match is better.